clippy_lints/src/matches.rs

```diff
@@ -849,63 +868,119 @@ fn check_single_match_opt_like<'a>(
+        for &(ty_path, pat_path) in CANDIDATES {
             if path == pat_path && match_type(cx, ty, ty_path) {
                 return true;
             }
         }
         false
-    };
-    if paths_and_types.iter().all(in_candidate_enum) {
-        report_single_match_single_pattern(cx, ex, arms, expr, els);
+    }
+
+    /// Returns true if the given pattern contains only one of `CANDIDATES` enums that could be
+    /// simplified by the `matches` lint.
+    pub fn is_known_enum(cx: &LateContext<'_>, pat: &Pat<'_>) -> bool {
+        match pat.kind {
+            PatKind::Path(ref path) => contains(cx, &qpath_to_string(path), cx.typeck_results().pat_ty(pat)),
+            PatKind::TupleStruct(ref path, ..) => {
+                let ty = cx.typeck_results().pat_ty(pat);
+                let path = &qpath_to_string(path);
+                (contains_only_wilds(pat) || contains_single_binding(pat)) && contains(cx, path, ty)
+            },
+            _ => false,
+        }
     }
 }
 
-/// Collects paths and their types from the given patterns. Returns true if the given pattern could
-/// be simplified, false otherwise.
-fn collect_pat_paths<'a>(acc: &mut Vec<(String, Ty<'a>)>, cx: &LateContext<'a>, pat: &Pat<'_>, ty: Ty<'a>) -> bool {
-    match pat.kind {
-        PatKind::Wild => true,
-        PatKind::Tuple(inner, _) => inner.iter().all(|p| {
-            let p_ty = cx.typeck_results().pat_ty(p);
-            collect_pat_paths(acc, cx, p, p_ty)
-        }),
-        PatKind::TupleStruct(ref path, ..) => {
-            let path = rustc_hir_pretty::to_string(rustc_hir_pretty::NO_ANN, |s| {
-                s.print_qpath(path, false);
-            });
-            acc.push((path, ty));
-            true
+/// Returns true if the given arms of the pattern forms exhaustive match that can be simplified to
+/// `if let` construction.
+fn check_exhaustive<'a>(cx: &LateContext<'a>, left: &Pat<'_>, right: &Pat<'_>, ty: Ty<'a>) -> bool {
+    match (&left.kind, &right.kind) {
+        (PatKind::Wild, _) | (_, PatKind::Wild) => true,
+        (PatKind::Tuple(left_in, left_pos), PatKind::Tuple(right_in, right_pos)) => {
+            check_exhaustive_tuples(cx, left_in, left_pos, right_in, right_pos)
         },
-        PatKind::Binding(BindingAnnotation::Unannotated, .., ident, None) => {
-            acc.push((ident.to_string(), ty));
-            true
+        (
+            PatKind::TupleStruct(left_qpath, left_in, left_pos),
+            PatKind::TupleStruct(right_qpath, right_in, right_pos),
+        ) => {
+            let are_structs_with_the_same_name =
+                || -> bool { cx.qpath_res(left_qpath, left.hir_id) == cx.qpath_res(right_qpath, right.hir_id) };
+            let are_known_enums =
+                || -> bool { known_enums::is_known_enum(cx, left) && known_enums::is_known_enum(cx, right) };
+            if are_structs_with_the_same_name() || are_known_enums() {
```

Comment:
Just check the type on one side. We can assume the types are equal.

Reply:
Could you please explain in more detail?

We actually check here if we have structures with the same name. This is required to correctly handle the following case (added as unit test at [7fc3545](https://github.com/rust-lang/rust-clippy/pull/8395/commits/7fc3545cd6eae7d94aff3b2e07d3317be8b2cae7)):
```rust
    enum E {
        S1(i32, i32),
        S2(i32, i32),
    }
    let s = E::S1(1, 2);

    // don't lint
    match s {
        E::S1(..) => {},
        E::S2(..) => {},
    }
```

We detect these cases using `are_structs_with_the_same_name`. This check should be performed for `TupleStructs` patterns only, because any other case will cause compile error. For example, we cannot match for `Struct` patterns with the different types.

The second condition `are_known_enums` is required to check if we have one of the standard `TupleStruct` patterns in the both arms. This allows us to handle cases like this:

```rust
match x {
  // `Err` and `Ok` are both `TupleStruct`'s, so we must process them here.
  Err(..) => {},
  Ok(..) => {},
}
```

Reply:
In your first example the type of both patterns is `E`. In the second example the type of both patterns is `Result`. The type is always the same. The name of the enum variant or the enum doesn't matter since you can just check the type.

Reply:
> But we want to make sure that the both variants have the same type as well.

This doesn't make sense. You are checking the type of the pattern and the type of the pattern is the enum, not the variant.

I disagree that it should lint if the variants are the same. That is still an exhuastive match of the enum. The point of #8282 is that the `match` expression will no longer compile if a variant is added to the enum, and that is what the user wants because the compiler helps them update their code.

Reply:
Well, my idea was is that the user may want to extend the enum with more variants later. So the `match` expression will give him warnings about non-exhaustive matching.

But suggesting to replace such `match` to `if let` is reasonable and should not be considered as a false positive. So, I agree, let's lint these patterns as well.

Reply:
> The case we want to lint is if the enum has a variant with the single struct, and we match this struct in both arms:

> ```rust
>     // lint
>     match s {
>         E::S1(42, ..) => {},
>         E::S1(..) => {},
>     }
> ```

> I disagree that it should lint if the variants are the same.

> So, I agree, let's lint these patterns as well.

Our communication broke down somewhere. I think we should not lint this case.

Reply:
Ah, sorry for that. These corner cases are not easy to understand.

You mean, that we should not lint cases with non-standard enums on both arms at all? Even if they have a single member.

e.g.:
```rust
   // don't lint
    match s {
        E::S1(42, ..) => {},
        E::S1(..) => {},
    }

    // lint, because we have a wild pattern
    match s {
        E::S1(42, ..) => {},
        _ => {},
    }
```

And we could simplify the condition for the lint case just to `is_known_enum(left_tuple_pattern)`, right? We could do this, because if the first arm have one of the standard types (`Result`, `Option`, etc.), we can be sure, that the second one have the same type or wildcard/binding pattern.

So the condition will become:
```rust
        (PatKind::TupleStruct(_, left_in, left_pos), PatKind::TupleStruct(_, right_in, right_pos))
            if contains_only_wilds(right) =>
        {
            if !contains_only_known_enums(cx, &left) {
                return false;
            }
            check_exhaustive_tuples(cx, left_in, left_pos, right_in, right_pos)
        },
```

Reply:
That sounds right 👍